### ISSUE TITLE
Functions integer-to-string and string-to-integer with radix

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#241]
I propose (in response to a suggestion from Joel Kalvesmaki on the XML Slack) two functions

```
fn:integer-to-string($value as xs:integer, $radix as xs:integer := 10) as xs:string

fn:string-to-integer($value as xs:string, $radix as xs:integer := 10) as xs:integer
```

that produce/accept string representations of signed integers in a base other than 10. Proposed range for $radix is 2 to 32.

Note:  I considered extending `fn:format-integer()`, but the result looked clumsy.

**Comment by ndw** [comment]
Yes, it can. Done. :-)

**Comment by ChristianGruen** [comment]
+1. Convenient functions of this kind are indeed missing in the Binary Module, which is why we also defined custom functions similar to the ones proposed in this issue (https://docs.basex.org/wiki/Conversion_Module#Numbers).

**Comment by Arithmeticus** [comment]
Thank you @michaelhkay for the concise, elegant approach to making hex functionality more convenient.

One might need a 3-arity version as well, something like `... $digit-picture as xs:string` to allow users to define the mapping of integers along the lines of `fn:translate()`, where the default is `0123456789ABCDEFGHIJ....` 

Rationale: the hexavigesimal system I'm most familiar with assigns A to 0, B to 1, etc. Base-64 binary also begins this way.

Bonus: one could map Devengari or Eastern Arabic numeration systems.

What to do when `$radix` is 100, 1000, or 1000000000?

**Comment by benibela** [comment]
Radix could also be detected from the string rather than defaulting to 10.  Like prefix `0x` being hexadecimal. `$` also being hexadecimal (it is a Pascal notation). (but not prefix `0` for octal, that gives nothing but confusion)

I made such functions, too: https://www.benibela.de/documentation/internettools/xpath-functions.html#x-integer

**Comment by michaelhkay** [comment]
I suggest we also consider allowing hex integer literals using the familiar 0xNNNN syntax (useful, for example, in codepoints-to-string()); and while we're about it, allow numeric literals to include "_" separators for readability.

**Comment by line-o** [comment]
I believe these two function and the additional features discussed here are very much needed additions to the core of XQuery. Thanks for this elegant approach.

**Comment by michaelhkay** [comment]
For reading hex (or other) numbers, I suggest a new function `parse-integer($value as xs:string, $radix as xs:integer  := 10)`. There's a lot of detail needed on exactly what's allowed, e.g. leading signs, grouping separators, etc.

For writing hex (or other) numbers, I suggest adding an optional `$radix` argument to `format-integer()` (the alternative would be to add more micro-syntax to the picture string). Again there's a lot of detail needed, for example how negative numbers are handled.

For literals, I suggest we extend the rules for numeric literals (a) to allow hex integers in the form 0xFFFFF, and (b) to allow embedded underscores for readability.

**Comment by joewiz** [comment]
I look forward to writing an update to https://joewiz.org/2021/05/30/converting-roman-numerals-with-xquery-xslt/ saying that the 6-line Closure parser for Roman numerals and 24-line XQuery 3.1 adaptation can be replaced with a simple XPath 4.0 function call: `parse-integer("XLIV", "I")`!

**Comment by line-o** [comment]
@michaelhkay should `fn:string-to-integer` be replaced by `fn:parse-integer`?

**Comment by ChristianGruen** [comment]
I’m wondering if  the function name isn’t confusing. `parse-integer('1')` is obvious, but for `parse-integer('a', 16)` , `string-to-integer`, `parse-to-integer`, or `convert-to-integer` could be more intuitive. But we already have lots of `parse` functions in the spec, and some more were added just recently (`parse-ietf-date`, `parse-QName`, `parse-uri`).

> I look forward to writing an update to https://joewiz.org/2021/05/30/converting-roman-numerals-with-xquery-xslt/ saying that the 6-line Closure parser for Roman numerals and 24-line XQuery 3.1 adaptation can be replaced with a simple XPath 4.0 function call: `parse-integer("XLIV", "I")`!

;·) And we could add support for:

```
parse-integer('quatre-vingt-dix-neuf', $lang := 'fr')`,
parse-integer('六十五万四千三百二十一', $lang := 'ja')`,
parse-integer('𒐦𒐠𒐜𒐐𒐌', $lang := 'sux'),
...

…and…

```xquery
parse-decimal('50°30'15'', $format := 'sexagesimal')
```

**Comment by Arithmeticus** [comment]
The examples suggest that a 3-arity version of the function will be needed, with the 3rd parameter being a map. Possible map members:

- `$are-digits as xs:boolean := true()`

If true:
- `$digit-convention as xs:string := 'indic'` _others: roman, grc-acrophonic, grc-alphabetic, etc. Native support for indic (limited pool of digits that cycle across units, tens, hundreds, etc.), but implementation dependent for others? There are different methods of Roman numeration systems, and are eligible to be synthesized with others, e.g., some Greek acrophonic conventions._
- `$digit-sequence as xs:string+ := ('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')` _requires as many members as the value of $radix_

If false:
- `$lang as xs:string?` _lang code, implementation-dependent_

_A language code for a digit system makes no sense to me. The script is detectable from Unicode attributes and one language tradition may have many digit systems._

**Comment by michaelhkay** [comment]
My feeling is that the requirement to handle hexadecimal numbers is for applications in the world of computer technology, and that use of western digits is ubiquitous in that application domain. This isn't about handling ancient Babylonian inscriptions with a number base of 60; it's about handling HTML color codes and the like. I see no reason to handle digits other than 0-9, a-f.

**Comment by Arithmeticus** [comment]
I think any _de minimis_ approach must also support every data type that is already central to the specs. Which means that `xs:base64Binary` should be supported, which entails an enlargement of the alphabet, `[a-zA-Z\d+/=]`, and a different calculus.

**Comment by Arithmeticus** [comment]
I retract that suggestion. If one has a string representation of base64Binary, it can/should be parsed with `xs:base64Binary()` then changed to some other numeric later, depending on how it should  be interpreted.

**Comment by michaelhkay** [comment]
I embarked on writing a PR to support hex literals, and hit a stumbling block: should they be signed or unsigned? In Java and similar languages, 0xFF represents +255, while 0xFFFFFFFF represents -1. That doesn't work well for XPath where integers are of arbitrary size rather than fixed at 32 bits. It's hard to see how we could avoid surprises here; if we make hex literals look like Java, people will assume 32-bit twos complement.

In consequence, I'm inclined to support hex-to-integer conversion only as a function, where an additional parameter can be used to indicate whether it's signed or unsigned.

But that doesn't totally solve the problem. I can see how to provide a function parse-integer(value, radix) that handles signed integers in any radix, or a function parse-hex-integer(value, is-signed) that handles twos-complement signed values, but I can't see how to provide both capabilities in one function.

**Comment by ChristianGruen** [comment]
I think we’ll be pretty safe if we always return positive numbers unless the input string is not prefixed by a minus sign, and allow an error if the result exceeds an implementation-defined integer limit. Java and XPath integers have never shared the same limits, and Java triggers an error, too, if the integer limit is exceeded.

This would also be consistent with various other programming languages. Here are some calls that I just tried:

Language | Code | Result
--- | --- | ---
Perl | `hex("FFFFFFFF")` | 4294967295
Python | `int("FFFFFFFF",16)` | 4294967295
PHP | `hexdec("FFFFFFFF")` | 4294967295
JavaScript | `parseInt('FFFFFFFF', 16)` | 4294967295
Java | `Long.parseLong("FFFFFFFF", 16)` | 4294967295
Java | `Integer.parseInt("FFFFFFFF", 16)` | NumberFormatException: For input string: "FFFFFFFF"

If we included an `is-signed` flag, I believe it should be numeric (8, 16, 32, …; i.e., not limited to 32 bits) and available for other radixes as well (mainly 2, possibly also 4, 8, and 32). It may be sufficient, though, to add examples in the documentation that demonstrate how an unsigned value can be converted to an unsigned one:

```xquery
let $limit := 256
let $value := parse-integer('11111111', 2) (: 255 :)
return if($value >= $limit div 2) then (
  $value - $limit
) else (
  $value
)
```

**Comment by ChristianGruen** [comment]
> Proposed range for $radix is 2 to 32.

I would suggest extending the maximum radix value to `36`. I remember a project requirement of ours where integer IDs needed to be compactified by converting them to any combination of numbers and digits (`36` → `10`, `46655` → `ZZZ`, etc.).

Next, 36 seems to be used as a maximum value in several other programming languages.

**Comment by ndw** [comment]
I agree with @ChristianGruen that we should support radixes (radii, gah, I'm an insufficiently experienced educated American to know which of these is greek and which is latin and when it matters) from 2-36.

**Comment by ndw** [comment]
Closed by #434
